Add webpage for Generic Table support (#1889)
apache/polaris: site/content/in-dev/unreleased/generic-table.md

# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#
title: Generic Table (Beta)
type: docs
weight: 435
---

The Generic Table in Apache Polaris provides basic management support for non-Iceberg tables. 

With the Generic Table API, you can:
- Create generic tables under a namespace
- Load a generic table 
- Drop a generic table
- List all generic tables under a namespace

**NOTE** The current generic table is in beta release, there can still be incomplete features and bugs. Please use it
with caution and report any issue if encountered.

## What is a Generic Table?

A generic table in Polaris is a structured entity that defines the necessary information. Each 
generic table entity contains the following properties:

- **name** (required): A unique identifier for the table within a namespace
- **format** (required): The format for the generic table, i.e. "delta", "csv"
- **base-location** (optional): Table base location in URI format. For example: s3://<my-bucket>/path/to/table
  - The table base location is a location that includes all files for the table

>   - A table with multiple disjoint locations (i.e. containing files that are outside the configured base location) is not compliant with the current generic table support in Polaris.

Author: Do you mean volume usage? I think how volume is going to be supported in Polaris is currently not discussed yet. Since this is a beta feature, if we decided to support the use cases with mutilple locations, we can evolve quickly to support this.

  - If no location is provided, clients or users are responsible for managing the location.

> - **properties** (optional): Properties for the generic table passed on creation

Reviewer: Does the Spark plugin use any specific property names? If yes, it would be good to add a section for them:
* as an illustration,
* to avoid name clashes with other use cases.

Author: Do you mean does spark plugin actually do transfer of specific property it receives from Spark and convert to a reserved key name ? Today, we do not do this. All properties a defined by Spark, and Spark have the right to update any of the property keys, which i don't think it is a good idea to doc it here for Generic tables

Reviewer: I'm afraid, I'm a bit lost here :sweat_smile: 

I mean: do we know what properties are currently set/retrieved from this `properties` list by Polaris code?

Reviewer: Suppose a Generic Table named `A` exists and has some properties usable from Spark via our new Spark Client. Now, if another client wants to query the same table via the Generic Tables API, how should that (new) client interpret existing table properties?

The new client may not be aware of the Spark Client, but the properties are observable via API. I believe this aspect of the Generic Tables API ought to be clarified in this doc.

Reviewer: True, but the REST catalog spec is not owned / defined by Polaris, while this one is.

I do not insist on a complete enumeration of possibilities. I think it should be sufficient to make a broad statement, for example: at this time, the contents of the `properties` map are not strictly defined and their interpretation is delegated to client / engine implementations, including interoperability concerns.

In fact, this is what happens with Iceberg tables too, IMHO, but I believe it is valuable to be explicit in specs.

Reviewer: thx - sgtm

- **doc** (optional): Comment or description for the table

## Generic Table API Vs. Iceberg Table API

Generic Table provides a different set of APIs to operate on the Generic Table entities while Iceberg APIs operates on
the Iceberg Table entities.

> | Operations   | **Generic Table API**                                       | **Iceberg Table API**                                       |

Reviewer: What is the Operation for making change to the data in a Generic Table?

Reviewer 2: The client does that, not the server

Reviewer: Do you mean that the server does not "know" what such a table change? If so, it certainly deserves a dedicated paragraph :sweat_smile:  As for me, I tend to view the Generic Tables API as something similar to the Iceberg REST Catalog API, which does control commits and by extension conflict resolution on the server side.

|--------------|-------------------------------------------------------------|-------------------------------------------------------------|

> | Create Table | create an Iceberg table                                     | create a generic table                                      |

Reviewer: Are titles of columns 2 and 3 swapped WRT row contents by any chance?

Author: Yes, sorry for the mistake.

Reviewer: thx for the fix :+1:

> | Load Table   | load an Iceberg table, load a generic table throws an error | load a generic table, load an Iceberg table throws an error |

Reviewer: Does this mean that attempting to load an Iceberg table via the Generic Tables API will throw an error?

Reviewer: Please consider explaining this in more details. As for me I had to read this table a few times to try and understand the intended behaviour :sweat_smile:

Author: yes

Author: added more description, hopefully it is more clear now

> | Drop Table   | drop an Iceberg table, drop a generic table throws an error | drop a generic table, drop an Iceberg table throws an error |

Reviewer 2: I think the  note here should be like "drop a generic table, **dropping** an Iceberg table throws an error"

Author: updated. i also uses `.` in between to avoid potential confusion to user

| List Table   | list all Iceberg tables                                     | list all generic tables                                     |

Note that generic table shares the same namespace with Iceberg tables, the table name has to be unique under the same namespace.

## Working with Generic Table

There are two ways to work with Polaris Generic Tables today:

> 1) Directly communicate with Polaris through REST API calls using curl. Details will be described in the later section.

Reviewer: `using curl` seems to be an example rather than a requirement here, right? I suppose any tool capable to making HTTP requests will work just as well.

Author: Good point, i updated the wording to `using tools such as curl`

2) Use the Spark Client provided if you are working with Spark. Please refer to [Polaris Spark Client]({{% ref "polaris-spark-client" %}}) for detailed instructions.

### Create a Generic Table

To create a generic table, you need to provide the corresponding fields as described in [What is a Generic Table](#what-is-a-generic-table).
Following is the REST API for creating a generic Table:

```json

> POST /polaris/v1/{prefix}/namespaces/{namespace}/generic-tables

Reviewer: nit: since this block of text is in JSON format the `POST` line should probably be outside (otherwise it violates the format)

Author: Thanks for pointing that out, move it out to shell format

Reviewer: pint: the format issues is still there :)

{
  "name": "<table_name>",
  "format": "<table_format>",
  "base-location": "<table_base_location>",
  "doc": "<comment or description for table>",
  "properties": {
    "<property-key>": "<property-value>"
  }
}
```

Here is an example to create a generic table with name `delta_table` and format as `delta` under a namespace `delta_ns` 
for catalog `delta_catalog`:

```json
POST /polaris/v1/delta_catalog/namespaces/delta_ns/generic-tables
{
  "name": "delta_table",
  "format": "delta",
}
```

### Load a Generic Table
The REST API for load a generic table is the following:

```json
GET /polaris/v1/{prefix}/namespaces/{namespace}/generic-tables/{generic-table}
```

Here is an example to load the table `delta_table`:
```json
GET /polaris/v1/delta_catalog/namespaces/delta_ns/generic-tables/{generic-table}
```
And the response looks like the following:
```json
{
  "table": {
    "name": "delta_table",
    "format": "delta",
    "base-location": null,
    "doc": null,

>     "properties": null

Reviewer: This example is rather confusing IMHO. It provides neither the location nor properties... What does it represent?

Reviewer 2: This looks like a minimal generic table, with no optional fields set? But it's not a great example, since the table created in the above section had optional fields populated.

Reviewer: This example shows that there is a Delta table named `delta_table`, but what can the reader/user do with this information other than avoid creating other tables with the same name?

  }
}
```

### List a Generic Table
Here is the REST API for listing the generic tables under a given namespace:
```json
GET /polaris/v1/{prefix}/namespaces/{namespace}/generic-tables/
```

Following rest call lists all tables under namespace delta_namespace:
```json
GET /polaris/v1/delta_catalog/namespaces/delta_ns/generic-tables/
```
Example Response:
```json
{
  "identifiers": [
    {
      "namespace": ["delta_ns"],
      "name": "delta_table"
    }
  ],
  "next-page-token": null
}
```

### Drop a Generic Table
The drop generic table REST API is the following:
```json
DELETE /polaris/v1/{prefix}/namespaces/{namespace}/generic-tables/{generic-table}
```

To drop the table `delat_table`, use the following:
```json
DELETE /polaris/v1/delta_catalog/namespaces//generic-tables/{generic-table}
```

### API Reference

> For the complete and up-to-date API specification, see the [generic-tables-api.yaml](https://github.com/apache/polaris/blob/main/spec/polaris-catalog-apis/generic-tables-api.yaml).

Reviewer: nit: maybe use a `swagger.io` reference as in #1879 ?

Reviewer: Ideally this should point to the same YAML version as the version of the doc (e.g. 1.0.0 vs. `main`).... not sure how to do it, though :thinking:

Author: updated the link to the swagger.io, but seems we only have the catalog bundle yaml, so updated the text to Catalog API Spec also.

## Limitations

> The support for cross engine sharing of Generic Table is very limited:

Author: updated

1) Limited spec information. Currently, there is no spec for information like Schema, Partition etc. 

> 2) No commit coordination or update capability provided at the catalog service level.

Reviewer 3: To me this is a very serious issue. No way to coordinate changes means there will be consistency issues.

Reviewer 4: Not all formats even have a way to do transactional commits. The basic premises here is to behave like the Spark Catalog with HMS (or Unity) which have these same guarantees for any sources. 

For example registering a Cassandra would work but there is nothing in the Polaris world that would (or could) manage commits for a C* source.  

Another example would be Delta Lake, which only can optionally (in 4.0) use a Catalog based commit coordinator and usually does not even consider the catalog when making commits. 

Polaris is only guaranteeing a consistent view of the metadata about the entity, not any guarantees about the underlying data.

Reviewer 4: A user is tying a table-identifier to a bucket of properties which an engine can use to resolve and read a table.

For example, Let's say I want to make a Delta Lake table in Spark with Unity, or HMS, or whatnot. The Delta catalog plugin will essentially bundle up a set of properties that say Table X is at location Y and potentially has other properties that should be used when accessing this table (Split sizes, compute behaviors etc ...) . That is packed up and put in a Catalog be in HMS or Unity or whatnot. A user when using SQL can reference the Identifier and the Catalog Plugin within Spark will pull down the set properties and use them to initialize a Spark Format Specific plugin to use those properties. In Delta's case it will load up the Delta Lake plugin and use the properties in the Catalog as the Datasource Reader properties. The datasource will then load a Spark Table object and return it to the user (or use it in a query). In the case of Delta the only properties actually relating to the table's state are the "Location" and "format", everything else is essentially re-loaded from the delta log found at the Location.

If instead you had a Cassandra entity, it would have properties format="cassandra" and then a bunch of properties which would dictate connection parameters for the C* cluster. Actual information about the "state" of the table other than it's connection format can only be read by loading the appropriate plugin and using the properties. 

This is model used by Spark for interacting with all Spark sources that don't have an independent V2 Catalog implementation (and actually works for those with a V2 Catalog Implementation as well as long as they also have a V2 source register implementation). 

So the user benefit of Polaris providing this is essentially allowing Polaris to be used as a generic Spark catalog and gaining support for all DSv1 and most DSv2 sources without having an additional store for a Spark cluster.

Reviewer 2: > So if basically only the table name is there, what's a user's benefit for it?

I'm confused -- are we cross-examining the feature here or documenting it?

The benefit is as is documented here; you can use Delta and other non-Iceberg tables in Spark using the Spark connector. The doc walks you through how that works.

Author: ```
So if basically only the table name is there, what's a user's benefit for it?
```
As @RussellSpitzer mentioned, the milestone polaris accomplishes today is enable polaris as a centralized catalog service for Spark Catalog. Furthermore, for delta, the state is inside the delta log, as far as the client is able to load the delta log, "base-location" is the only information needed to enable access to the delta table. I can try to make it more clear in the doc.

Reviewer 3: What I really do not like is implying promises (`A data catalog serves as the central registry for a table’s metadata. It manages transactions and table state, as well as access controls and read/write interoperability.`) which just do not hold true: there is no way to manage transactions and table state, control access, etc.

WRT to what's in Polaris - it's incomplete (I suspect there's no doubt there).

Reviewer 4: I don't actually think it's incomplete in that context? I wouldn't imagine we would ever support those things for this endpoint and I don't think it would be a surprise to any user of Spark who uses this?

Reviewer: I agree with @snazy 's point that the blog post (link above) is not really aligned with the Generic Tables API as far as Catalog as a "central registry" for table's metadata is concerned. The Generic Table API actually diminishes the role of the catalog as metadata registry by delegating most of the metadata loading to the client (even `location` is optional in the API).

That said, as far as this PR is concerned, I believe it should be sufficient to describe actual behaviour of Polaris in this respect. There is certainly room for improvements in terms of clarity and precision in this doc, but I think the current state of this PR is probably acceptable for 1.0.

> Today, all those responsibility will be at the client side. For example, the Delta log serialization, deserialization and update are all done at Client side.

Reviewer 2: ```suggestion
Today, all of these responsibilities are on the client side. For example, Delta log serialization, deserialization and updates are all done on the Client side.
```

Author: updated

---
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#
title: Generic Table (Beta)
type: docs
weight: 435
---

The Generic Table in Apache Polaris is designed to provide support for non-Iceberg tables across different table formats includes delta, csv etc. It currently provides the following capabilities:
- Create a generic table under a namespace

> - Load a generic table 

Reviewer 2: The API itself is actually named loadGenericTable, so I don't think it's exactly misleading. It does load the generic table's metadata, which is whatever metadata was registered in the catalog during createGenericTable

> - Load a generic table 

Author: yeah, i was mainly try to matching the term of other systems to use "load or get", and I don't think load has to be specifically to the iceberg metadata.

Note that generic table shares the same namespace with Iceberg tables, the table name has to be unique under the same namespace.

## Working with Generic Table

There are two ways to work with Polaris Generic Tables today:
1) Directly communicate with Polaris through REST API calls using tools such as `curl`. Details will be described in the later section.
2) Use the Spark client provided if you are working with Spark. Please refer to [Polaris Spark Client]({{% ref "polaris-spark-client" %}}) for detailed instructions.

### Create a Generic Table

To create a generic table, you need to provide the corresponding fields as described in [What is a Generic Table](#what-is-a-generic-table).
Following is the REST API for creating a generic Table:

```shell
POST /polaris/v1/{prefix}/namespaces/{namespace}/generic-tables
```
```json
{
  "name": "<table_name>",
  "format": "<table_format>",
  "base-location": "<table_base_location>",
  "doc": "<comment or description for table>",
  "properties": {
    "<property-key>": "<property-value>"
  }
}
```

> Here is an example to create a generic table with name `delta_table` and format as `delta` under a namespace `delta_ns` 

Reviewer: Since there's no "update" API, what happens if there's a mistake in the initial create request? WIll the client have to delete and re-create the table?

I do not mean to cause API changes at this point, just trying to clarity things for potential non-Polaris readers.

    "doc": "delta table example",
    "properties": {
      "key1": "value1"
    }
  }
}
```

### List a Generic Table
Here is the REST API for listing the generic tables under a given namespace:
```shell
GET /polaris/v1/{prefix}/namespaces/{namespace}/generic-tables/
```

Following rest call lists all tables under namespace delta_namespace:
```shell
GET /polaris/v1/delta_catalog/namespaces/delta_ns/generic-tables/
```
Example Response:
```json
{
  "identifiers": [
    {
      "namespace": ["delta_ns"],
      "name": "delta_table"
    }
  ],
  "next-page-token": null
}
```

### Drop a Generic Table
The drop generic table REST API is the following:
```shell
DELETE /polaris/v1/{prefix}/namespaces/{namespace}/generic-tables/{generic-table}
```

To drop the table `delat_table`, use the following:
```shell
DELETE /polaris/v1/delta_catalog/namespaces/delta_ns/generic-tables/{generic-table}
```

### API Reference

For the complete and up-to-date API specification, see the [Catalog API Spec](https://editor-next.swagger.io/?url=https://raw.githubusercontent.com/apache/polaris/refs/heads/main/spec/generated/bundled-polaris-catalog-service.yaml).

## Limitations

> The Generic Table support today is very limited:

Reviewer 5: Can we reword a bit like this? 
```suggestion
Current limitations of Generic Table support:
```
